File: src/gsocialsim/stimuli/data_source.py
from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional
import csv

from gsocialsim.stimuli.stimulus import Stimulus, MediaType


class DataSource(ABC):
    """Abstract base class for any source of external data."""

    @abstractmethod
    def get_stimuli(self, tick: int) -> List[Stimulus]:
        """Returns a list of all stimuli that should be injected at a given tick."""
        raise NotImplementedError


def _clean_opt_str(v: Any) -> Optional[str]:
    if v is None:
        return None
    if isinstance(v, str):
        v = v.strip()
        return v if v else None
    # defensive: coerce non-strings to string
    s = str(v).strip()
    return s if s else None


class CsvDataSource(DataSource):
    """
    A concrete data source that reads from a CSV file.

    Required columns:
      - id
      - tick
      - source
      - content_text

    Optional columns:
      - topic
      - stance           (float in [-1, 1])
      - identity_threat  (float in [0, 1])
      - political_salience (float in [0, 1])
      - primal_triggers  (csv list: e.g. "self,contrast,visual")
      - primal_intensity (float in [0, 1])
      - media_type       (news, social_post, video, meme, longform, forum_thread)
      - creator_id       (for subscription targeting)
      - outlet_id        (for subscription targeting)
      - community_id     (for subscription targeting)

    Notes:
      - All optional fields are safe: missing columns are simply treated as None.
      - The Stimulus class also mirrors some values into stimulus.metadata to keep the
        CSV format flexible and backward compatible.
    """

    def __init__(self, file_path: str):
        self.stimuli_by_tick: Dict[int, List[Stimulus]] = {}

        with open(file_path, mode="r", encoding="utf-8") as infile:
            reader = csv.DictReader(infile)

            # Normalize header names defensively (CSV authors make mistakes).
            # We keep original keys but also allow case-insensitive matching.
            fieldnames = reader.fieldnames or []
            lower_to_actual = {fn.lower(): fn for fn in fieldnames}

            def get(row: Dict[str, Any], key: str) -> Any:
                actual = lower_to_actual.get(key.lower(), key)
                return row.get(actual)

            for row in reader:
                tick_raw = get(row, "tick")
                if tick_raw is None:
                    continue
                tick = int(tick_raw)

                stimulus_id = str(get(row, "id"))
                source = str(get(row, "source"))
                content_text = str(get(row, "content_text"))

                topic = _clean_opt_str(get(row, "topic"))
                media_type_raw = _clean_opt_str(get(row, "media_type"))
                stance_raw = get(row, "stance")
                stance_val = None
                if stance_raw is not None and str(stance_raw).strip() != "":
                    try:
                        stance_val = float(stance_raw)
                    except Exception:
                        stance_val = None
                threat_raw = get(row, "identity_threat")
                threat_val = None
                if threat_raw is not None and str(threat_raw).strip() != "":
                    try:
                        threat_val = float(threat_raw)
                    except Exception:
                        threat_val = None
                pol_raw = get(row, "political_salience")
                pol_val = None
                if pol_raw is not None and str(pol_raw).strip() != "":
                    try:
                        pol_val = float(pol_raw)
                    except Exception:
                        pol_val = None
                primal_raw = get(row, "primal_triggers")
                primal_triggers = None
                if primal_raw is not None and str(primal_raw).strip() != "":
                    tokens = [t.strip().lower() for t in str(primal_raw).replace("|", ",").split(",")]
                    primal_triggers = [t for t in tokens if t]
                primal_intensity_raw = get(row, "primal_intensity")
                primal_intensity_val = None
                if primal_intensity_raw is not None and str(primal_intensity_raw).strip() != "":
                    try:
                        primal_intensity_val = float(primal_intensity_raw)
                    except Exception:
                        primal_intensity_val = None
                creator_id = _clean_opt_str(get(row, "creator_id"))
                outlet_id = _clean_opt_str(get(row, "outlet_id"))
                community_id = _clean_opt_str(get(row, "community_id"))

                # Keep metadata flexible and compatible with existing code paths.
                metadata: Dict[str, Any] = {"topic": topic}
                if stance_val is not None:
                    metadata["stance"] = stance_val
                if threat_val is not None:
                    metadata["identity_threat"] = threat_val
                if pol_val is not None:
                    metadata["political_salience"] = pol_val
                if primal_triggers is not None:
                    metadata["primal_triggers"] = primal_triggers
                if primal_intensity_val is not None:
                    metadata["primal_intensity"] = primal_intensity_val
                if media_type_raw is not None:
                    metadata["media_type"] = media_type_raw
                if creator_id is not None:
                    metadata["creator_id"] = creator_id
                if outlet_id is not None:
                    metadata["outlet_id"] = outlet_id
                if community_id is not None:
                    metadata["community_id"] = community_id

                stimulus = Stimulus(
                    id=stimulus_id,
                    source=source,
                    tick=tick,
                    content_text=content_text,
                    media_type=MediaType.from_any(media_type_raw),
                    creator_id=creator_id,
                    outlet_id=outlet_id,
                    community_id=community_id,
                    topic_hint=topic,
                    stance_hint=stance_val,
                    political_salience=pol_val,
                    primal_triggers=primal_triggers,
                    primal_intensity=primal_intensity_val,
                    metadata=metadata,
                )

                self.stimuli_by_tick.setdefault(tick, []).append(stimulus)

        print(
            f"Loaded {sum(len(s) for s in self.stimuli_by_tick.values())} stimuli from {file_path}"
        )

    def get_stimuli(self, tick: int) -> List[Stimulus]:
        return self.stimuli_by_tick.get(tick, [])

File: cpp/src/data_source.cpp
#include "data_source.h"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <sstream>

#include "utils.h"

static std::string trim_copy(const std::string& s) {
    size_t start = 0;
    while (start < s.size() && std::isspace(static_cast<unsigned char>(s[start]))) {
        ++start;
    }
    size_t end = s.size();
    while (end > start && std::isspace(static_cast<unsigned char>(s[end - 1]))) {
        --end;
    }
    return s.substr(start, end - start);
}

static std::string to_lower_copy(const std::string& s) {
    std::string out = s;
    for (char& c : out) {
        if (c >= 'A' && c <= 'Z') {
            c = static_cast<char>(c - 'A' + 'a');
        }
    }
    return out;
}

static std::optional<std::string> clean_opt_str(const std::string& v) {
    std::string t = trim_copy(v);
    if (t.empty()) return std::nullopt;
    return t;
}

static std::optional<double> parse_opt_double(const std::string& v) {
    std::string t = trim_copy(v);
    if (t.empty()) return std::nullopt;
    try {
        return std::stod(t);
    } catch (...) {
        return std::nullopt;
    }
}

CsvDataSource::CsvDataSource(const std::string& file_path) {
    std::ifstream infile(file_path);
    if (!infile.is_open()) {
        return;
    }

    std::string header_line;
    if (!std::getline(infile, header_line)) {
        return;
    }

    std::vector<std::string> headers = split_csv_row(header_line);
    std::unordered_map<std::string, size_t> lower_to_idx;
    for (size_t i = 0; i < headers.size(); ++i) {
        lower_to_idx[to_lower_copy(trim_copy(headers[i]))] = i;
    }

    auto get_field = [&](const std::vector<std::string>& row, const std::string& key) -> std::string {
        auto it = lower_to_idx.find(to_lower_copy(key));
        if (it == lower_to_idx.end()) return "";
        if (it->second >= row.size()) return "";
        return row[it->second];
    };

    std::string line;
    while (std::getline(infile, line)) {
        if (line.empty()) continue;
        auto row = split_csv_row(line);

        auto tick_raw = get_field(row, "tick");
        auto tick_opt = parse_opt_double(tick_raw);
        if (!tick_opt.has_value()) continue;
        int tick = static_cast<int>(tick_opt.value());

        std::string id = trim_copy(get_field(row, "id"));
        std::string source = trim_copy(get_field(row, "source"));
        std::string content_text = trim_copy(get_field(row, "content_text"));
        if (id.empty() || source.empty()) {
            continue;
        }

        auto topic = clean_opt_str(get_field(row, "topic"));
        auto media_type_raw = clean_opt_str(get_field(row, "media_type"));
        auto stance_val = parse_opt_double(get_field(row, "stance"));
        auto threat_val = parse_opt_double(get_field(row, "identity_threat"));
        auto pol_val = parse_opt_double(get_field(row, "political_salience"));

        std::vector<std::string> primal_triggers;
        auto primal_raw = clean_opt_str(get_field(row, "primal_triggers"));
        if (primal_raw.has_value()) {
            std::string raw = primal_raw.value();
            std::replace(raw.begin(), raw.end(), '|', ',');
            std::stringstream ss(raw);
            std::string token;
            while (std::getline(ss, token, ',')) {
                auto t = to_lower_copy(trim_copy(token));
                if (!t.empty()) {
                    primal_triggers.push_back(t);
                }
            }
        }

        auto primal_intensity_val = parse_opt_double(get_field(row, "primal_intensity"));
        auto creator_id = clean_opt_str(get_field(row, "creator_id"));
        auto outlet_id = clean_opt_str(get_field(row, "outlet_id"));
        auto community_id = clean_opt_str(get_field(row, "community_id"));

        Stimulus stim;
        stim.id = id;
        stim.source = source;
        stim.tick = tick;
        stim.content_text = content_text;
        stim.media_type = media_type_from_string(media_type_raw.value_or(""));
        stim.creator_id = creator_id;
        stim.outlet_id = outlet_id;
        stim.community_id = community_id;
        stim.topic_hint = topic;
        stim.stance_hint = stance_val;
        stim.political_salience = pol_val;
        stim.primal_triggers = primal_triggers;
        stim.primal_intensity = primal_intensity_val;

        if (topic.has_value()) stim.metadata["topic"] = topic.value();
        if (stance_val.has_value()) stim.metadata["stance"] = std::to_string(stance_val.value());
        if (threat_val.has_value()) stim.metadata["identity_threat"] = std::to_string(threat_val.value());
        if (pol_val.has_value()) stim.metadata["political_salience"] = std::to_string(pol_val.value());
        if (!primal_triggers.empty()) {
            std::string joined;
            for (size_t i = 0; i < primal_triggers.size(); ++i) {
                if (i) joined += ",";
                joined += primal_triggers[i];
            }
            stim.metadata["primal_triggers"] = joined;
        }
        if (primal_intensity_val.has_value()) {
            stim.metadata["primal_intensity"] = std::to_string(primal_intensity_val.value());
        }
        if (media_type_raw.has_value()) stim.metadata["media_type"] = media_type_raw.value();
        if (creator_id.has_value()) stim.metadata["creator_id"] = creator_id.value();
        if (outlet_id.has_value()) stim.metadata["outlet_id"] = outlet_id.value();
        if (community_id.has_value()) stim.metadata["community_id"] = community_id.value();

        stimuli_by_tick_[tick].push_back(stim);
    }
}

std::vector<Stimulus> CsvDataSource::get_stimuli(int tick) {
    auto it = stimuli_by_tick_.find(tick);
    if (it == stimuli_by_tick_.end()) return {};
    return it->second;
}
